griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,293 @@
+        content = params["values"]["content"]
+        filename = params["values"]["path"]
+
+        try:
+            self._save_to_drive(filename, content)
+
+            return InfoArtifact(f"saved successfully")
+        except Exception as e:
+            return ErrorArtifact(f"error saving file to Google Drive: {e}")
+
+    def _save_to_drive(self, filename: str, value: any, parent_folder_id=None) -> InfoArtifact | ErrorArtifact:
+        service = self._build_client(self.UPLOAD_FILE_SCOPES)
+    
+        if isinstance(value, str):
+            value = value.encode()
+    
+        file_metadata = {"name": filename}
+        if parent_folder_id:
+            file_metadata["parents"] = [parent_folder_id]
+    
+        media = MediaIoBaseUpload(BytesIO(value), mimetype="application/octet-stream", resumable=True)
+    
+        try:
+            file = service.files().create(body=file_metadata, media_body=media, fields="id").execute()
+            return InfoArtifact(file.get("id"))
+        except HttpError as error:
+            return ErrorArtifact(f"An error occurred: {error}")
+
+    @activity(config={
+        "description": "Download a file from Google Drive based on a provided path",
+        "schema": Schema({
+            Literal("file_path",
+                    description="Path of the Google Drive file (like 'MainFolder/Subfolder1/filename.ext') "
+                                "that needs to be downloaded."): str
+        })
+    })
+    def download_file(self, params: dict) -> BlobArtifact | ErrorArtifact:
+        values = params["values"]
+
+        try:
+            service = self._build_client(self.LIST_FILES_SCOPES)
+            file_id = self._path_to_file_id(service, values["file_path"])
+            if not file_id:
+                return ErrorArtifact(f"Could not find file: {values['file_path']}")
+
+            file_info = service.files().get(fileId=file_id).execute()
+            mime_type = file_info['mimeType']
+
+            if mime_type in self.GOOGLE_EXPORT_MIME_MAPPING:
+                export_mime = self.GOOGLE_EXPORT_MIME_MAPPING[mime_type]
+                request = service.files().export_media(fileId=file_id, mimeType=export_mime)
+            else:
+                request = service.files().get_media(fileId=file_id)
+
+            downloaded_file = request.execute()
+            logging.info(f"File '{values['file_path']}' successfully downloaded.")
+            return BlobArtifact(downloaded_file)
+
+        except HttpError as e:
+            logging.error(e)
+            return ErrorArtifact(f"error downloading file from Google Drive: {e}")
+        except MalformedError:
+            logging.error("MalformedError occurred")
+            return ErrorArtifact("error downloading file due to malformed credentials")
+
+    @activity(config={
+        "description": "Search for a file on Google Drive based on name or content",
```

Comment:
Should read as "Can be used to search for files..."

Reply:
Done!